DOCS-ConnectedAppsRevamp OICD and OAUTH 2.0 (#76)
mulesoft/docs-access-management: modules/ROOT/pages/connected-apps-overview.adoc

Anypoint Platform supports OAuth 2.0 and OpenID Connect to authorize apps to access data within Anypoint Platform.
OAuth (Open Authorization) is an open protocol that provides secure API authorization from applications in a simple and standardized way.
OAuth 2.0 can authorize access to resources without revealing user credentials to apps.
OpenID Connect identifies the end user and obtains information to pass to OAuth 2.0 connected apps.

> // Sara add some indentation or mark up here please

Reviewer: Let's keep comments out of docs and use the Github comments feature for communication

> The Connected Apps feature uses OAuth 2.0 and OpenID Connect, or _OIDC_, as part of its authentication process. OAuth 2.0 and OpenID Connect flows must be completed in order to grant access to Anypoint APIs only to services and users that have the proper permissions. 

Reviewer: ```suggestion
The Connected Apps feature implements OAuth 2.0 and OpenID Connect, or _OIDC_, for authorizing access to data. OAuth 2.0 and OpenID Connect flows grant permission to services and users to access the Anypoint APIs.
```

Reviewer: I'm trying to figure out what to do with that second sentence 
- Oauth is only ever user->app, never user->user
- The flow _is what grants_ the permission. Phrasing makes it seem like you need the flow and the permission

Author: @maxwellgerber Is that better?

=== OAuth 2.0

When understanding OAuth 2.0, an authorization protocol, we must first define several key terms to ground our understanding. The first is the client, which refers to the resource that wants the data and could be a computer or server. Next, the authorization server refers to the server that can authorize, and the resource server refers to the system with the data. The access token is the key to get into data in the resource server and the redirect URI is where you go if the user or service authorizes the other service.

> There are two types of flows for OAuth 2.0. 

Reviewer: There are _many_ OAuth 2.0 flows. These are the two most common.

> There are two types of flows for OAuth 2.0. 

Author: @sarathecoubian Can you change this to read, what Max said, I still don't understand how to edit.

`authorization_code`:: 
In the authorization_code flow, the client navigates to the authorization server. The server then requests authorization for specific scopes, or _permissions_, from the user. If the user authorizes the client to have this information, an authorization code is sent the redirect URI. The redirect URI then exchanges the authorization code for an access token from the authorization server. The access token is then sent to the resource server to obtain the data by means of a specific API. This flow demonstrates _an application acting on behalf of the user_.

`client_credentials`::
In the client_credentials flow, the client navigates to the authorization server. The server then requests authorization for specific scopes, or _permissions_, from the application. If the application authorizes the client to have this information, an authorization code is sent the redirect URI. The redirect URI then exchanges the authorization code for an access token from the authorization server. The access token is then sent to the resource server to obtain the data by talking to a specific API. This flow demonstrates a _service-to-service_ use case.

=== OpenID Connect (OIDC)

> OICD is an authentication protocol that authenticates users. This identity layer sits on top of the OAuth 2.0 protocol and provides authentication as opposed to the authorization OAuth 2.0 provides.

Reviewer: OIDC extends and complements OAuth - it doesn't oppose it.

Anypoint Platform supports OAuth 2.0 and OpenID Connect to authorize apps to access data within Anypoint Platform.
OAuth (Open Authorization) is an open protocol that provides secure API authorization from applications in a simple and standardized way.
OAuth 2.0 can authorize access to resources without revealing user credentials to apps.
OpenID Connect identifies the end user and obtains information to pass to OAuth 2.0 connected apps.

The Connected Apps feature implements OAuth 2.0 and OpenID Connect, or _OIDC_, for authorizing access to data. OAuth 2.0 and OpenID Connect flows grant permission to services and users to access the Anypoint APIs.

=== OAuth 2.0

The OAuth 2.0 authorization protocol depends on several elements to function. 

Client::

> The resource that requests data on behalf of a user or service.

Reviewer: Refer to https://tools.ietf.org/html/rfc6749#section-1.1 for a precise definition of a client. A client is _not_ a resource.

> The resource that requests data on behalf of a user or service.

Author: ```suggestion
An application that makes requests for data on behalf of a user or service with authorization.
```

> Authorization server::

Reviewer: @sarathecoubian should the second word be capitalized? Here and elsewhere
```suggestion
Authorization Server::
```

Reviewer 2: Nah. No title casing for definition lists, to my knowledge.

> The server that gives access tokens to the client once it does authorization.

Reviewer: what is `it`? The server or the client? [avoid ambiguous pronouns](https://developers.google.com/style/pronouns).

> The server that gives access tokens to the client once it does authorization.

Author: ```suggestion
The server that gives access tokens to the client once the server does authorization.
```

> Resource server::

Author: ```suggestion
Resource Server::
```

The system that contains the requested data.

Access token:: 

> A "key" needed to access data in the resource server. 

Reviewer: `Access Keys` are [another method of authz](https://swagger.io/docs/specification/authentication/api-keys/) that predates OAuth & is generally regarded as not best practice. Can you find a different way to describe tokens that differentiates them from keys? (The RFC is helpful here too)

> A "key" needed to access data in the resource server. 

Author: ```suggestion
A token needed to access data from the resource server. 
```

Redirect URI::
The destination the authorization server sends the user or server once the application is authorized.

> There are many types of OAuth 2.0 flows; the Connected Apps feature uses the following:

Reviewer: We use more than just these two - we support all openid flows + refresh token + password too

Author: @maxwellgerber Which flows specifically do we support?

> There are many types of OAuth 2.0 flows; the Connected Apps feature uses the following:

Author: ```suggestion
There are many types of OAuth 2.0 flows; the Connected Apps feature most commonly uses the following:
```

`authorization_code`:: 
In the authorization_code flow, the client navigates to the authorization server. The server then requests authorization for specific scopes, or _permissions_, from the user. If the user authorizes the client to have this information, an authorization code is sent the redirect URI. The redirect URI then exchanges the authorization code for an access token from the authorization server. The access token is then sent to the resource server to obtain the data by means of a specific API. This flow demonstrates _an application acting on behalf of the user_.

`client_credentials`::
In the client_credentials flow, the client navigates to the authorization server. The server then requests authorization for specific scopes, or _permissions_, from the application. If the application authorizes the client to have this information, an authorization code is sent the redirect URI. The redirect URI then exchanges the authorization code for an access token from the authorization server. The access token is then sent to the resource server to obtain the data by talking to a specific API. This flow demonstrates a _service-to-service_ use case.

=== OpenID Connect (OIDC)

> OICD is an authentication protocol that authenticates users. This identity layer complements OAuth 2.0 protocol and provides authentication at a layer above it.

Reviewer 2: ```suggestion
OIDC is an authentication protocol that authenticates users. This identity layer complements OAuth 2.0 protocol and provides authentication at a layer above it.
```
```suggestion
OICD is an authentication protocol that authenticates users. This identity layer complements OAuth 2.0 protocol and provides authentication at a layer above it.
```

Anypoint Platform supports OAuth 2.0 and OpenID Connect to authorize apps to access data within Anypoint Platform.
OAuth (Open Authorization) is an open protocol that provides secure API authorization from applications in a simple and standardized way.
OAuth 2.0 can authorize access to resources without revealing user credentials to apps.
OpenID Connect identifies the end user and obtains information to pass to OAuth 2.0 connected apps.

> // Sara add some indentation or mark up here please

Author: ```suggestion
```

The Connected Apps feature uses OAuth 2.0 and OpenID Connect, or _OIDC_, as part of its authentication process. OAuth 2.0 and OpenID Connect flows must be completed in order to grant access to Anypoint APIs only to services and users that have the proper permissions. 

=== OAuth 2.0

The OAuth 2.0 authorization protocol depends on several elements to function. 

Client::
The resource that requests data on behalf of a user or service.

Authorization server::
The server that gives access tokens to the client once it does authorization.

Resource server::
The system that contains the requested data.

Access token:: 
A "key" needed to access data in the resource server. 

Redirect URI::
The destination the authorization server sends the user or server once the application is authorized.

> There are two types of flows for OAuth 2.0. 

Author: ```suggestion
There are many types of OAuth 2.0 flows; Connected Apps uses the following:
```

> OAuth can authorize access to resources without revealing user credentials to apps.

Reviewer 2: ```suggestion
OAuth 2.0 can authorize access to resources without revealing user credentials to apps.
```

OpenID Connect identifies the end user and obtains information to pass to OAuth 2.0 connected apps.

// Sara add some indentation or mark up here please

> In order to understand the purpose of Connected Apps, there are several technical concepts and terms to be familiar with, beginning with OAuth 2.0 (commonly referred to as just OAuth) and OpenID Connect (commonly referred to as OICD). Before breaking down these terms, it’s important to understand that these flows are completed in order to give access to Anypoint APIs to only those services and users with the proper permissions. 

Reviewer 2: This seems less about Connected Apps directly and more about the authentication aspect of it. I'll be running under this assumption herein.

Reviewer 2: ```suggestion
The Connected Apps feature uses OAuth 2.0 and OpenID Connect, or _OIDC_, as part of its authentication process. OAuth 2.0 and OpenID Connect flows must be completed in order to grant access to Anypoint APIs only to services and users that have the proper permissions. 
```

Reviewer 2: We do not want to refer to OAuth 2.0 as OAuth for short. OAuth 2.0 is a completely revamped version of OAuth, so I'll be scouting this out in docs and fixing it. I think this is wrong in a few legacy docs too.

> OAuth 2.0

Reviewer 2: ```suggestion
=== OAuth 2.0

```

> When understanding OAuth 2.0, an authorization protocol, we must first define several key terms to ground our understanding. The first is the client, which refers to the resource that wants the data and could be a computer or server. Next, the authorization server refers to the server that can authorize, and the resource server refers to the system with the data. The access token is the key to get into data in the resource server and the redirect URI is where you go if the user or service authorizes the other service.

Reviewer 2: No first person in docs. Additionally, we cannot lead the users through a narrative and tell them what to understand; we simply provide the information. This is a great writing convention for blogs and more casual material, but technical documentation has to be very tightly woven and not tell them what to think; all it can do is state what is.

Reviewer 2: ```suggestion
The OAuth 2.0 authorization protocol depends on several elements to function. 

Client::
The resource that requests data on behalf of a user or service.

Authorization server::
The server that gives access tokens to the client once it does authorization.

Resource server::
The system that contains the requested data.

Access token:: 
A "key" needed to access data in the resource server. 

Redirect URI::
The destination the authorization server sends the user or server once the application is authorized.
```

Reviewer 2: This prints as a definition list. So now...

- re: Client: Request data from whom?
- re: Auth server: Authorize what/who?
- re: Redirect URI: what other service?

Author: @sebastiankorol Do you have good answers to these? I'm not sure how to clarify.

Reviewer 3: You can refer to [RFC 6749](https://tools.ietf.org/html/rfc6749) for a precise definition of these terms

Author: I tried to put these in my own words based on Seba's resource

Client: The resource that requests data on behalf of a user or service.
Auth server: The server that gives access tokens to the client once it does authorization
Redirect URI: The destination the authorization server sends the user or server once the application is authorized. 

Also feel free to link the link Seba provided

Author: Updated this, so Let me know

Reviewer: Did you `git push` the changes?

There are two types of flows for OAuth 2.0. 

> The first is the authorization_code flow. In this flow, the client goes to the authorization server and and the server pops up a window asking for authorization for specific scopes from the user. If the user authorizes the client to have this information, an authorization code is sent the redirect URI. The redirect URI then exchanges the authorization code for an access token from the authorization server which is then taken to the resource server to obtain the data by talking to a specific API. This is called an application acting on behalf of the user.

Reviewer 2: ```suggestion
`authorization_code`:: 
In the authorization_code flow, the client navigates to the authorization server. The server then requests authorization for specific scopes, or _permissions_, from the user. If the user authorizes the client to have this information, an authorization code is sent the redirect URI. The redirect URI then exchanges the authorization code for an access token from the authorization server. The access token is then sent to the resource server to obtain the data by means of a specific API. This flow demonstrates _an application acting on behalf of the user_.
```

Reviewer 2: Do we have alternative phrasing for "talking to" an API?

Author: Is that better?

> The second flow is the client_credentials flow. In this flow, each piece is very similar but instead of a user authorizing the application, the application can authorize on its own behalf. This is also called a service-to-service use case.

Reviewer 2: ```suggestion
`client_credentials`::
In the client_credentials flow, the client navigates to the authorization server. The server then requests authorization for specific scopes, or _permissions_, from the application. If the application authorizes the client to have this information, an authorization code is sent the redirect URI. The redirect URI then exchanges the authorization code for an access token from the authorization server. The access token is then sent to the resource server to obtain the data by talking to a specific API. This flow demonstrates a _service-to-service_ use case.
```

Reviewer 2: Better to restate everything than rely on legacy content in case something is deprecated. I doubt that would happen here, but it is good practice.

> OpenID Connect (OICD)

Reviewer 2: ```suggestion
=== OpenID Connect (OIDC)

```

> OICD is an authentication protocol that authenticates users. This identity layer sits on top of the OAuth 2.0 protocol and provides authentication as opposed to the authorization OAuth 2.0 provides.

Reviewer 2: ```suggestion
OICD is an authentication protocol that authenticates users. This identity layer relies on the OAuth 2.0 protocol and provides authentication as opposed to the authorization that OAuth 2.0 provides.
```

> OICD is an authentication protocol that authenticates users. This identity layer sits on top of the OAuth 2.0 protocol and provides authentication as opposed to the authorization OAuth 2.0 provides.

Reviewer 2: Is there a way to rephrase that first sentence? Also, the second part of this is dangerously close to other published materials on OIDC. We have to be careful not to plagiarize.

Author: Not sure how to update it but could we say: 

OICD is an authentication protocol that validates users. This identity layer works with the OAuth 2.0 protocol and provides authentication as opposed to the authorization OAuth 2.0 provides

> OICD is an authentication protocol that authenticates users. This identity layer sits on top of the OAuth 2.0 protocol and provides authentication as opposed to the authorization OAuth 2.0 provides.

Author: OICD is an authentication protocol that validates users. This identity layer works with the OAuth 2.0 protocol and provides authentication as opposed to the authorization OAuth 2.0 provides
